### PR TITLE
feat: only build versions that dont exist on dist

### LINKED ISSUE #260: Building without downloading everything
We should be rebuilding dists without downloading _everything_.

### PR DESCRIPTION
**Pull request by olizilla**
This PR reworks things to only build binaries that aren't already published on dist.ipfs.io. The releases dir now only needs to contain the files that are to change, and is treated as a patch to apply to the exist root CID for dist.ipfs.io.

I assert that this will work and be rad because content-addressing.

Also of note, this is a someewhat of a pre-requisite for #273 so we can start building dist in CI, and get binaries signed without requiring everyone to own apple hardware. On the default branch building dist takes about 25GB of disk and about 2hrs on a Quad-Core i7. It doesn't need to if we trust the hashes.

TODO:
- [x] only build what changed
- [x] fix website generation to not depend on a complete local releases dir
- [x] fix publishing to patch new release into existing dist.ipfs.io CID
- [x] Move the build step to CI, rather than expecting a maintainer to artisanally compile the bins on their machine

For subsequent PRs
- Add mac notarizing in CI as per #273 
- Push the diff to cluster
- Update the DNSLink from CI
- AUTOMATE ALL TEH THINGZ!!!1111!
  - Post go-ipfs binaries to the release from CI!?
  - watch for new semver tag on go-ipfs repo and re-build automatically!?

Fixes #260 

License: MIT
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

File: build-go.sh
Comment (olizilla) [inline build-go.sh:306]: The list of versions that need building is now calculated during the build, so we change this to expect a list rather than a file name.

File: build-go.sh
Comment (olizilla) [inline build-go.sh:331]: I should add a check to see if `ipfs` exists and the deamon is online.

File: build-go.sh
Comment (olizilla) [inline build-go.sh:392]: Moved the _copy the versions file_ step here so we only copy it if it changed.

File: package.json
Comment (olizilla) [inline package.json:7]: extracted to it's own script so we can run `npm run standard -- --fix` and have it auto fix the lints.

File: package.json
Comment (olizilla) [inline package.json:9]: the `dist` target adds data files to `site/data/releases` but `clean` wasn't cleaning it, so now it does.

File: scripts/dist.js
Comment (olizilla) [reply scripts/dist.js:9]: and the build process will now only have dist.json files for the dists that are chaning, (typically 1 at time) so it pulls the others from the existing dist.ipfs.io... it's pulling them from the gateway, but I could update this to use `ipfs-http-client` and the local IPFS node as is done in `./scripts/patch.js`

**Comment by olizilla** [comment]
I know the dist build process could be alot nicer. This is first adventure in there, so my goal here was not to change any more than necessary to get patch style build/publish working. I over-stepped that, but not by much, the core of the build process is left unchanged.

@aschmahmann we should step through this together when you have time.
@dignifiedquire if you have a moment to sanity check this then that would be rad. If not, no worries!

File: scripts/dist.js
Comment (olizilla) [reply scripts/dist.js:9]: - ✅ Updated to use ipfs rather than the gateway to fetch published dist.json

**Review by dignifiedquire — APPROVED** [review]
looks sane to me 👍

File: build-go.sh
Comment (olizilla) [reply build-go.sh:331]: Added a check to deps-check.sh and updated it so we check it early and halt the build if any are missing in https://github.com/ipfs/distributions/pull/320/commits/3398a234ce493f8fa780a3873e2d5b222e1bbfee

**Comment by olizilla** [comment]
@lidel do these changes bring you joy?

**Review by lidel — CHANGES_REQUESTED** [review]
Removing the need for fetching entire thing brings joy! :heart: :raised_hands: 

But I think we need to make it more self-contained and reproducible.

Found some issues while trying to run `make publish` locally (fresh `git clone`)

## a bug while fetching `dist.json`

Without it `JSON.parse` got invalid input and failed like this:

```
> dist.ipfs.io@1.0.0 build /home/lidel/project/ipfs/distributions
> npm run dist && npm run build:assets && npm run build:hugo


> dist.ipfs.io@1.0.0 dist /home/lidel/project/ipfs/distributions
> node scripts/dist.js

Updating hugo data for dist website
- Fetching dist.json for fs-repo-migrations v1.6.4 from /ipns/dist.ipfs.io
- Fetching dist.json for go-ipfs v0.7.0 from /ipns/dist.ipfs.io
- Fetching dist.json for gx v0.14.2 from /ipns/dist.ipfs.io
- Fetching dist.json for gx-go v1.9.0 from /ipns/dist.ipfs.io
- Fetching dist.json for ipfs-cluster-ctl v0.13.0 from /ipns/dist.ipfs.io
SyntaxError: Unexpected token , in JSON at position 3670
    at JSON.parse (<anonymous>)
    at fetchDistData (/home/lidel/project/ipfs/distributions/scripts/dist.js:44:21)
```

Perhaps we need `concat` instead of `all`?

```diff
-const all = require('it-all')
+const concat = require('it-concat')
-    jsonStr = await all(ipfs.cat(`${DIST_ROOT}/${distName}/${version}/dist.json`))
+    jsonStr = await concat(ipfs.cat(`${DIST_ROOT}/${distName}/${version}/dist.json`), { type: 'string' })
```

## Missing menu?

Unsure if I am doing something wrong, but `make publish` created `QmPvuHWJb6dpeZ6GwWGQuEPBSUcLHqCtLq2ALdRaqwn62j` which is missing menu:

> Screenshot_2020-10-28 IPFS Distributions

**Comment by olizilla** [comment]
Many thank yous for your time @lidel I shall fix!

**Comment by olizilla** [comment]
@lidel I switched the dist.js to use IPFS instead of fetching from the gateway and broke it! **and** I added in more cleaning which broke the site build. Is fixed now!

Screenshot 2020-10-28 at 17 53 17

> But I think we need to make it more self-contained and reproducible.

Strong agree. This is my _how the heck does dist work_ PR. Next up is reproducibuild++

**Comment by olizilla** [comment]
Dist now builds in in a github workflow in 1min 23s! ✨🎷🐩
https://github.com/ipfs/distributions/pull/320/checks?check_run_id=1322838224

**Review by lidel — APPROVED** [review]
I re-checked to see if it will pick up `v0.8.0-rc1` added upstream and it worked as expected (detected upstream data and skipped/reused it):

```console
$ ./dist.sh add-version go-ipfs v0.8.0-rc1
WARNING: not marking pre-release go-ipfs v0.8.0-rc1 as the current version.
../../build-go.sh "go-ipfs" "github.com/ipfs/go-ipfs" "cmd/ipfs" versions
comparing versions with /ipns/dist.ipfs.io/go-ipfs/versions
skipping go-ipfs - all versions published at /ipns/dist.ipfs.i
```

I suggest we give it a try: merge this PR and then see if adding `v0.8.0-rc2` works as expected. 
That will be a safe way to test if it behaves as expected without impact on real release.

**Comment by jessicaschilling** [comment]
@olizilla, do we have plans for wrapping this up or should it be iceboxed?

**Comment by olizilla** [comment]
@aschmahmann pls may i have your official blessing to merge this and unblock #327

**Comment by aschmahmann** [comment]
@olizilla I was hoping to test this with RC2 (probably next week 🤞) just to verify that all the outputs look as expected (i.e. build before + after the PR to verify things look right).

If you've already done that though then we can probably just merge and still check the diffs on publish to verify things look right

**Comment by olizilla** [comment]
I verified it at the time of writing, and @lidel tested it out. I'm very happy to wait for your test run too tho, as it's a significant change.

**Review by aschmahmann — APPROVED** [review]
This is amazing! A huge thank you to @olizilla and everyone who committed, reviewed and tested this PR.

**Pull request merged**
